engine/uState.ml

```diff
@@ -1030,11 +1030,29 @@ let demote_seff_univs univs uctx =
 
 let demote_global_univs env uctx =
   let env_ugraph = Environ.universes env in
-  let global_univs = UGraph.domain env_ugraph in
-  let global_constraints, _ = UGraph.constraints_of_universes env_ugraph in
-  let promoted_uctx =
-    ContextSet.(of_set global_univs |> add_constraints global_constraints) in
-  { uctx with local = ContextSet.diff uctx.local promoted_uctx }
+  let (local_univs, local_constraints) = uctx.local in
+  let mem_univ u ugraph = match UGraph.check_declared_universes ugraph (Level.Set.singleton u) with
+  | () -> true
+  | exception (UGraph.UndeclaredLevel _) -> false
+  in
+  let mem_constraints (u, r, v) ugraph =
+    if mem_univ u ugraph && mem_univ v ugraph then match r with
+    | Le -> UGraph.check_leq ugraph (Universe.make u) (Universe.make v)
+    | Lt -> UGraph.check_leq ugraph (Universe.super @@ Universe.make u) (Universe.make v)
+    | Eq -> UGraph.check_eq_level ugraph u v
+    else false
```

Comment:
```suggestion
  let mem_constraint (u, _, v as cst) ugraph =
    mem_univ u ugraph && mem_univ v ugraph && UGraph.check_constraint ugraph cst
```

Reply:
With type-in-type that's not the same thing, but maybe we don't care.

Reply:
Am I missing something? AFAICT check_constraint, check_leq and check_eq_level all do `type_in_type || ...`

Reply:
Ah, you're write, I confused the UGraph API with the AcyclicGraph one.